### PR TITLE
Handle multiple env builds

### PR DESCRIPTION
**Pull request by joerowe**

### CODE REVIEW

File: microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java [hunk 1]
```diff
@@ -39,21 +43,22 @@ public static void main(String... args) {
...
     }
 
     private void build() {
         try {
-            long startTime = currentTimeMillis();
-            doBuild();
-            announce("\nGenerated [" + env + "] configs in " + (currentTimeMillis() - startTime) + "ms");
+            enableLogger(!jsonOutput);
+            doBuild(envs.contains("*")
+                    ? new MicroconfigRunner(rootDir, null).getMicroconfig().environments().environmentNames()
```
Comment (joerowe) [inline microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java:60]: would be nice if this was able to be grabbed without instantiating the runner, but i didnt want to mess with too much stuff !

File: microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java [hunk 2]
```diff
@@ -64,17 +69,19 @@ private void build() {
...
-        Properties properties = runner.buildProperties(env, groups, services);
-        if (jsonOutput) {
-            out.println(toJson(properties.save(asConfigResult())));
-        } else {
-            properties.save(runner.toFiles());
-        }
+    private void doBuild(Set<String> envs) {
+        envs.forEach(env -> {
+            long startTime = currentTimeMillis();
+            MicroconfigRunner runner = new MicroconfigRunner(rootDir, new File(String.format("%s/%s", destinationDir, env)));
```
Comment (alxmatorin) [inline microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java:75]: The feature from the pr is pretty good. But  there is a small problem. The current implementation breaks the backward compatibility.  If someone updates microconfig version and generates configs, he will expect to have generated configs in the same folder that used to be before the update. But after the pr, configs will be generated to {env} folder. Probably its better to add a new cli parameter(envs). if you use envs param - configs will be generated to the ${env} dir. if you use the old cli env param - configs will be generated to the destination dir without {env}

Comment (joerowe) [reply microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java]: done !

File: README.md [hunk 3]
```diff
@@ -958,7 +958,9 @@ To run the build you can download Microconfig release from https://github.com/mi
 
 The required build params:
 * `-r` - full or relative config root dir. 
-* `-e` - environment name (environment is used as a config profile, also as a group of services to build configs).
+* either `-e` or `-envs`.
+    * `-e` - environment name (environment is used as a config profile, also as a group of services to build configs).
+    * `-envs` - a comma separated list of environment names. Use "*" for all environments, and ! to exclude an environment. E.g. `-envs *, !base` to generate config for all environments _except_ `base`.
```
Comment (alxmatorin) [inline README.md:963]: param value must be without spaces. with space it doesnt work: -envs *,!base

File: utils/src/main/java/io/microconfig/utils/CollectionUtils.java [hunk 4]
```diff
@@ -43,4 +48,8 @@
     public static <T> Set<T> setOf(T... t) {
         return new LinkedHashSet<>(asList(t));
     }
+
+    public static boolean isCollectionEmpty(Collection c){
```
Comment (alxmatorin) [inline utils/src/main/java/io/microconfig/utils/CollectionUtils.java:52]: this method is not used anymore. you can rollback all changes from this class

File: microconfig-cli/src/test/java/io/microconfig/MicroconfigMainTest.java [hunk 5]
```diff
@@ -1,40 +1,110 @@
...
+
+    private void checkFileExist(String filePath, boolean shouldFileExist) {
+        File resultFile = new File(destinationDir, filePath);
+        assertEquals(shouldFileExist, resultFile.exists());
+        if(shouldFileExist) {
+            assertEquals("key: " + getName(resultFile), readFully(resultFile).trim());
+        }
+    }
+
+    @Ignore
```
Comment (alxmatorin) [inline microconfig-cli/src/test/java/io/microconfig/MicroconfigMainTest.java:105]: its better to remove this method

File: microconfig-cli/src/test/java/io/microconfig/MicroconfigMainTest.java [hunk 6]
```diff
@@ -1,40 +1,110 @@
 package io.microconfig;
 
+import jdk.nashorn.internal.ir.annotations.Ignore;
```
Comment (alxmatorin) [inline microconfig-cli/src/test/java/io/microconfig/MicroconfigMainTest.java:3]: wrong package

File: microconfig-cli/src/test/java/io/microconfig/MicroconfigParamsTest.java [hunk 7]
```diff
@@ -27,13 +32,21 @@ void rootDir() {
...
     }
 
     @Test
     void env() {
-        assertEquals("dev", params.env());
+        assertEquals("dev", params.environments().stream().findFirst().get());
+    }
+
+    @Test
+    void envs() {
```
Comment (alxmatorin) [inline microconfig-cli/src/test/java/io/microconfig/MicroconfigParamsTest.java:45]: can be simplified
```
@Test
    void envs() {
        assertEquals(new LinkedHashSet<>(asList("dev", "test")), params.environments());
    }

File: microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java [hunk 8]
```diff
@@ -39,42 +47,59 @@ public static void main(String... args) {
...
     }
 
     private void doBuild() {
-        enableLogger(!jsonOutput);
+        Set<String> envsToBuild = chooseEnvironments();
 
-        MicroconfigRunner runner = new MicroconfigRunner(rootDir, destinationDir);
+        envsToBuild.forEach(e -> {
+            long startTime = currentTimeMillis();
+            String filePath = envsToBuild.size()>1 ? String.format("%s/%s", destinationDir, e) : destinationDir;
```
Comment (alxmatorin) [inline microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java:78]: envsToBuild.size()>1 is wrong. cause you can run -envs prod or - envs * and you have only 1 env.

Comment (alxmatorin) [inline microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java:78]: I think (destinationDir + "/" + e) is esier to read then String.format("%s/%s", destinationDir, e)

File: microconfig-cli/src/test/java/io/microconfig/MicroconfigMainTest.java [hunk 9]
```diff
@@ -1,40 +1,110 @@
...
+        if(nestedDirectory!=null){
+            result = String.format("%s/%s", nestedDirectory, result);
+        }
+        return result;
+    }
+
+    private void checkFileExist(String filePath, boolean shouldFileExist) {
+        File resultFile = new File(destinationDir, filePath);
+        assertEquals(shouldFileExist, resultFile.exists());
+        if(shouldFileExist) {
```
Comment (alxmatorin) [inline microconfig-cli/src/test/java/io/microconfig/MicroconfigMainTest.java:100]: apply reformat to all changed code.
You can use default Idea autofomat style

File: microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java (hunk 8 above)
Comment (joerowe) [reply microconfig-cli/src/main/java/io/microconfig/MicroconfigMain.java:78]: agreed

File: microconfig-cli/src/main/java/io/microconfig/MicroconfigParams.java [hunk 10]
```diff
@@ -44,4 +44,23 @@ public boolean version() {
...
+            environments.add(env);
+        }
+        if (environments.isEmpty()) {
+            printErrorAndExit("set `-e (environment)` or `-envs (env1),(env2)...`");
+        }
+        return environments;
+    }
+
+    public boolean isSingleEnvBuild() {
+        return parser.value("e") != null && !parser.contains("envs");
```
Comment (alxmatorin) [inline microconfig-cli/src/main/java/io/microconfig/MicroconfigParams.java:64]: parser.contains("e") instead of parser.value("e") != null

**Comment by alxmatorin** [comment]
Looks good. Thank you for the contribution!

**Pull request merged**